ksqldb-engine/src/main/java/io/confluent/ksql/internal/StorageUtilizationMetrics.java

```diff
@@ -0,0 +1,254 @@
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.config.ConfigException;
+import org.apache.kafka.common.metrics.Gauge;
+import org.apache.kafka.common.metrics.KafkaMetric;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.MetricsContext;
+import org.apache.kafka.common.metrics.MetricsReporter;
+import org.apache.kafka.streams.StreamsConfig;
+
+public class StorageUtilizationMetrics implements MetricsReporter {
+  private static final String METRIC_GROUP = "ksqldb-utilization";
+
+  private final Map<String, Map<String, TaskStorageMetric>> metricsSeen;
+  private final Metrics metricRegistry;
+  static AtomicBoolean registeredNodeMetrics = new AtomicBoolean(false);
+
+  public StorageUtilizationMetrics() {
+    this(MetricCollectors.getMetrics());
+  }
+
+  public StorageUtilizationMetrics(final Metrics metricRegistry) {
+    this.metricsSeen = new HashMap<>();
+    this.metricRegistry = metricRegistry;
+  }
+
+  @Override
+  public void init(final List<KafkaMetric> list) {
+  }
+
+  @Override
+  public void configure(final Map<String, ?> map) {
+    final String dir;
+    if (map.containsKey(StreamsConfig.STATE_DIR_CONFIG)) {
+      dir = map.get(StreamsConfig.STATE_DIR_CONFIG).toString();
+    } else {
+      dir =  StreamsConfig
+        .configDef()
+        .defaultValues()
+        .get(StreamsConfig.STATE_DIR_CONFIG)
+        .toString();
+    }
+    configureShared(new File(dir), this.metricRegistry);
+  }
+
+  private static void configureShared(final File baseDir, final Metrics metricRegistry) {
+    if (registeredNodeMetrics.getAndSet(true)) {
+      return;
+    }
+    final MetricName nodeAvailable =
+        metricRegistry.metricName("node-storage-available", METRIC_GROUP);
+    final MetricName nodeTotal =
+        metricRegistry.metricName("node-storage-total", METRIC_GROUP);
+    final MetricName nodeUsed =
+        metricRegistry.metricName("node-storage-used", METRIC_GROUP);
+
+    metricRegistry.addMetric(
+        nodeAvailable,
+        (Gauge<Long>) (config, now) -> baseDir.getFreeSpace()
+    );
+    metricRegistry.addMetric(
+        nodeTotal,
+        (Gauge<Long>) (config, now) -> baseDir.getTotalSpace()
+    );
+    metricRegistry.addMetric(
+        nodeUsed,
+        (Gauge<Long>) (config, now) -> (baseDir.getTotalSpace() - baseDir.getFreeSpace())
+    );
+  }
+
+  @Override
+  public void metricChange(final KafkaMetric metric) {
+    if (!metric.metricName().name().equals("total-sst-files-size")) {
+      return;
+    }
+    final String taskId = metric.metricName().tags().getOrDefault("task-id", "");
+    final String queryId = getQueryId(metric);
+
+    // if we haven't seen a task for this query yet
+    synchronized (metricsSeen) {
```

Comment:
can you move this into another method like `private synchronized void handleNewSstFilesSizeMetric(final KafkaMetric metric, final String taskId, final String queryId) {...`?

ditto for the other side of this in `metricRemoval`

Reply:
And we have to do this because the hashmap is a concurrent hashmap, right? So when we access it we need to do it in a synchronized fashion? Is there something different about using a synchronized function rather than a synchronized block or is the synchronized function easier to use